### PR TITLE
[Assets] Generate checksum on asset save

### LINKED ISSUE #13362: [Improvement]: store a checksum for each asset
### Improvement description

There's value in knowing a checksum of a file, for example you can later have a feature which shows duplicates and allows the user to deduplicate in some way.

It probably makes sense it's MD5 just like `versions.binaryFileHash`, but would work even without versions being enabled. MD5 is also offered by cloud storages (for example, Azure blob storage automatically keeps an MD5 hash), some features in the future could benefit from that overlap, for example, "Check asset health" would go through the assets and check the cloud storage checksum against the one in the database to see they still match.

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #13362 

## Additional info  
- Added command `pimcore:assets:generate-checksums` to re-generate checksums

### CODE REVIEW

File: models/Asset.php [hunk 1]
```diff
@@ -1108,6 +1111,21 @@ public function getStream()
...
+    {
+        if (!$checksum = $this->getCustomSetting('checksum')) {
+            $this->generateChecksum();
+            $checksum = $this->getCustomSetting('checksum');
+        }
+
+        return $checksum;
+    }
+
+    public function generateChecksum(): void
```
Comment (brusch) [inline models/Asset.php:1127]:
```suggestion
    private function generateChecksum(): void
```
We should have this either `private` or `@internal`.

File: models/Asset.php [hunk 2]
```diff
@@ -1108,6 +1111,21 @@ public function getStream()
...
+    {
+        $this->setCustomSetting('checksum', Storage::get('asset')->checksum($this->getRealFullPath()));
```
Comment (brusch) [inline models/Asset.php:1129]:
```suggestion
        $this->setCustomSetting('checksum', Storage::get('asset')->checksum($this->getRealFullPath()));
        $this->getDao()->updateCustomSettings();
```
We could persist the info right away, right?

File: bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * @internal
+ */
+class GenerateChecksumCommand extends AbstractCommand
+{
+    protected function configure(): void
+    {
+        $this
+            ->setName('pimcore:assets:generate-checksums')
+            ->setDescription('Re-generates checksum for specific or all assets')
```
Comment (brusch) [inline bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php:35]: Should we maybe add an option to calculate the checksum for all assets that doesn't have one set?

**Review by brusch — CHANGES_REQUESTED** [review]

File: models/Asset.php (hunk 1 above)
Comment (dvesh3) [reply models/Asset.php:1127]: Marked internal

File: models/Asset.php (hunk 2 above)
Comment (dvesh3) [reply models/Asset.php:1129]: yes, done.

File: bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php (hunk 3 above)
Comment (dvesh3) [reply bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php:35]: Add option `missing-only` to generate checksum for assets not having one.

File: composer.json [hunk 4]
```diff
@@ -136,7 +136,8 @@
     "onnov/detect-encoding": "^2.0",
     "knplabs/knp-paginator-bundle": "^6.0.0",
     "masterminds/html5" : "^2.7",
-    "league/flysystem-bundle" : "^3.0",
+    "league/flysystem-bundle": "^3.1",
+    "league/flysystem": "^3.12.0",
```
Comment (brusch) [inline composer.json:140]: What's the reason for `^3.12.0` ? According to the [docs](https://flysystem.thephpleague.com/docs/usage/checksums/) checksums were added in `3.7.0`

**Review by brusch — CHANGES_REQUESTED** [review]

File: composer.json (hunk 4 above)
Comment (dvesh3) [reply composer.json:140]: @brusch that was also my first attempt to use `3.7.0` version but the tests were [failing](https://github.com/pimcore/pimcore/actions/runs/4604911044/jobs/8136327058). Also they fixed some dependencies in following [versions ](https://github.com/thephpleague/flysystem/blob/3.x/CHANGELOG.md#390---2022-10-18)so i just bumped the latest version.

Comment (brusch) [reply composer.json:140]: Ok, fine 😊👍

File: bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                        continue;
+                    }
+
+                    if ($asset->getType() === 'folder') {
+                        continue;
+                    }
+
+                    $this->output->writeln(' generating checksum for asset: ' . $asset->getRealFullPath() . ' | ' . $asset->getId());
+                    $asset->generateChecksum();
+                    $asset->save(['versionNote' => 'checksum generation']);
```
Comment (dpfaffenbauer) [inline bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php:112]: is there a need to call save when `generateChecksum` updates the custom settings anyway?

Comment (brusch) [reply bundles/CoreBundle/src/Command/Asset/GenerateChecksumCommand.php:112]: Good point - no - will fix that! 👍 Thanks a lot

**Comment by dkarlovi** [comment]
@dvesh3 thank you for working on this feature! \o/

**Pull request merged**